javalib/src/main/scala/java/util/AbstractCollection.scala

```diff
@@ -0,0 +1,89 @@
+package java.util
+
+import scala.annotation.tailrec
+import scala.collection.JavaConversions._
+import scala.collection.immutable.{List => SList}
+
+abstract class AbstractCollection[E] extends java.util.Collection[E] {
+  def iterator(): Iterator[E]
+  def size(): Int
+
+  @tailrec
+  private def iterateApplyAndDo[X](i: Iterator[E], f: (Iterator[E], E) => X, 
+      part: SList[X] = SList()): SList[X] = {
+    if (i.hasNext) iterateApplyAndDo(i, f, f(i, i.next()) :: part)
+    else part
+  }
+
+  @tailrec
+  private def iterateAndApply(i: Iterator[E], f: Iterator[E] => Unit): Unit =
+    if (i.hasNext) {
+      f(i)
+      iterateAndApply(i, f)
+    } else return
+
+  @tailrec
+  private def iterateCheckAndApply(i: Iterator[E], f: (Iterator[E], E) => Boolean): Boolean =
+    if (i.hasNext) {
+      if (f(i, i.next())) true
+      else iterateCheckAndApply(i, f)
+    } else false
+
+  def isEmpty(): Boolean = size == 0
+
+  def contains(o: Any): Boolean =
+    asScalaIterator(iterator).contains(o)
+  def containsAll(c: java.util.Collection[_]): Boolean =
+    asScalaIterator(c.iterator).foldLeft(false)((prev, item) => contains(item: Any) && prev)
+
+  def add(e: E): Boolean =
+    throw new UnsupportedOperationException()
+  def addAll[E1 <: E](c: java.util.Collection[E1]): Boolean =
+    throw new UnsupportedOperationException()
+
+  def clear(): Unit =
+    iterateAndApply(iterator, _.remove())
+
+  def remove(o: Any): Boolean = {
+    iterateCheckAndApply(iterator, {
+      case (i, e) if e.equals(o) =>
+        i.remove()
+        true
+      case _ => false
+    })
+  }
+
+  def removeAll(c: java.util.Collection[E]): Boolean = {
+    iterateApplyAndDo(iterator, {
+      case (i, e) if (c.contains(e)) => { i.remove(); true }
+      case _ => false
+    }).exists(_ == true)
+  }
+
+  def retainAll(c: java.util.Collection[E]): Boolean = {
+    iterateApplyAndDo(iterator, {
+      case (i, e) if (!c.contains(e)) => { i.remove(); true }
+      case _ => false
+    }).exists(_ == true)
+  }
+
+  override def toString(): String =
+    asScalaIterator(iterator).map(java.lang.String.valueOf(_)).mkString("[", ",", "]")
+
+  def toArray[X: reflect.ClassTag](a: Array[X]): Array[X] = {
+    val toFill: Array[X] = 
+      if (a.size >= size) a
+      else new Array(size)
+    val iter = iterator
+    for { i <- 0 until size } { toFill(i) = iter.next.asInstanceOf[X] }
+    for { i <- size until toFill.size } { toFill(i) = null.asInstanceOf[X] }
```

Comment:
a) Prefer `()` in `for` when that part is one-liner.
b) A side-effecting body (of the for) should always be on its own line.

So this would be written:
```scala
for (i <- 0 until size)
  toFill(i) = iter.next().asInstanceOf[X]
```